crates/uv/src/commands/project/remove.rs

```diff
@@ -0,0 +1,84 @@
+use anyhow::Result;
+use std::str::FromStr;
+use uv_distribution::pyproject_mut::PyProjectTomlMut;
+
+use distribution_types::IndexLocations;
+use pep508_rs::Requirement;
+use pypi_types::LenientRequirement;
+use uv_cache::Cache;
+use uv_configuration::{ExtrasSpecification, PreviewMode, Upgrade};
+use uv_distribution::ProjectWorkspace;
+use uv_warnings::warn_user;
+
+use crate::commands::{project, ExitStatus};
+use crate::printer::Printer;
+
+/// Remove one or more packages from the project requirements.
+#[allow(clippy::too_many_arguments)]
+pub(crate) async fn remove(
+    requirements: Vec<String>,
+    preview: PreviewMode,
+    cache: &Cache,
+    printer: Printer,
+) -> Result<ExitStatus> {
+    if preview.is_disabled() {
+        warn_user!("`uv remove` is experimental and may change without warning.");
+    }
+
+    // Find the project requirements.
+    let project = ProjectWorkspace::discover(&std::env::current_dir()?, None).await?;
+    let mut pyproject = PyProjectTomlMut::from_toml(project.current_project().pyproject_toml())?;
+
+    for req in requirements {
+        let req = Requirement::from(LenientRequirement::from_str(&req)?);
+        if pyproject.remove_dependency(&req)?.is_none() {
+            anyhow::bail!(
+                "the dependency `{}` could not be found in `dependencies`",
+                req.name
+            );
+        }
+    }
+
+    // Save the modified `pyproject.toml`.
+    fs_err::write(
+        project.current_project().root().join("pyproject.toml"),
+        pyproject.to_string(),
+    )?;
+
+    // Discover or create the virtual environment.
+    let venv = project::init_environment(&project, preview, cache, printer)?;
+
+    let index_locations = IndexLocations::default();
+    let upgrade = Upgrade::default();
+    let extras = ExtrasSpecification::default();
+    let exclude_newer = None;
+    let dev = false; // We only add regular dependencies currently.
```

Comment:
I'm not sure if this is correct? Should dev dependencies ever be synced even if they weren't modified?

Reply:
It's possible that the dev dependencies _could_ change by adding a non-dev dependency, because they're all locked together. (E.g., maybe the dep you add has a transitive dep on one of the dev deps, so now it's more constrained.) So, I guess we do have the include them? But it's not really right to sync them unconditionally. The same problem exists with extras. We shouldn't be syncing them unconditionally. We need to know which extras the user wants activated.

I'm wondering if `add` should imply `lock`, but not `sync`? Since locking is much simpler: we lock everything.

(I think there is a general problem here that exists as long as `add` implies `lock` or `sync`. (Rye has this problem too). Any arguments that are typically passed to `lock` or `sync` to guide resolution or installation now need to be passed to `add` too... So, e.g., we might need `exclude-newer` too.)